### PR TITLE
[NVPTX] Implement variadic functions using IR lowering

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch implements support for variadic functions for NVPTX targets.
The implementation here mainly follows what was done to implement it for
AMDGPU in https://github.com/llvm/llvm-project/pull/93362.

We change the NVPTX codegen to lower all variadic arguments to functions
by-value. This creates a flattened set of arguments that the IR lowering
pass converts into a struct with the proper alignment.

The behavior of this function was determined by iteratively checking
what the NVCC copmiler generates for its output. See examples like
https://godbolt.org/z/KavfTGY93. I have noted the main methods that
NVIDIA uses to lower variadic functions.

1. All arguments are passed in a pointer to aggregate.
2. The minimum alignment for a plain argument is 4 bytes.
3. Alignment is dictated by the underlying type
4. Structs are flattened and do not have their alignment changed.
5. NVPTX never passes any arguments indirectly, even very large ones.

This patch passes the tests in the `libc` project currently, including
support for `sprintf`.

### CODE REVIEW

File: clang/lib/CodeGen/Targets/NVPTX.cpp [hunk 1]
```diff
@@ -203,8 +203,15 @@ ABIArgInfo NVPTXABIInfo::classifyArgumentType(QualType Ty) const {
 void NVPTXABIInfo::computeInfo(CGFunctionInfo &FI) const {
   if (!getCXXABI().classifyReturnType(FI))
     FI.getReturnInfo() = classifyReturnType(FI.getReturnType());
-  for (auto &I : FI.arguments())
-    I.info = classifyArgumentType(I.type);
+
+  unsigned ArgumentsCount = 0;
+  for (auto &I : FI.arguments()) {
+    if (FI.isVariadic() && ArgumentsCount > 0)
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/Targets/NVPTX.cpp:209]: Special-casing argument 0 is sort of unusual; are you sure this is what you want?  Normally, I'd expect the distinguishing factor to be fixed vs. variadic: so all explicitly specified parameters are treated the same, then all arguments in the variadic list are passed the same.

Comment (jhuber6) [reply clang/lib/CodeGen/Targets/NVPTX.cpp:209]: You're right, this needs to account for all fixed arguments, not just the first (guaranteed) one. NVIDIA seems to handle it where the fixed arguments are passed using the regular ABI (can be indirect or direct) while the variadic arguments are always direct. Is there an easy way to check if an argument is part of the variadic set? Maybe if the argument number > number of arguments?

Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/NVPTX.cpp]: Look at amdgpu, it does exactly that normal lowering for fixed and something different for variadic.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp [hunk 2]
```diff
@@ -938,6 +938,37 @@ struct Amdgpu final : public VariadicABIInfo {
...
+  Value *initializeVaList(Module &M, LLVMContext &Ctx, IRBuilder<> &Builder,
+                          AllocaInst *, Value *Buffer) override {
+    return Builder.CreateAddrSpaceCast(Buffer, vaListParameterType(M));
+  }
+
+  VAArgSlotInfo slotInfo(const DataLayout &DL, Type *Parameter) override {
+    // NVPTX doesn't apply minimum alignment to types present in structs. Types
+    // with alignment less than four should be promoted by the compiler and will
+    // get the proper minimum alignment in those cases.
+    const unsigned MinAlign = 1;
```
Comment (JonChesterfield) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:964]: I don't think this does what the comment says it should. Also some half types might be passed with 2 byte alignment, others might be passed promoted to double.

**Review by JonChesterfield — COMMENTED** [review]
With the possible exception of some alignment handling this looks about as I'd expect it to. Ideally we'd get some feedback from nvptx-associated people but fixing libc is a good sign

File: llvm/lib/Target/NVPTX/NVPTXPassRegistry.def [hunk 3]
```diff
@@ -17,6 +17,8 @@
 #define MODULE_PASS(NAME, CREATE_PASS)
 #endif
 MODULE_PASS("generic-to-nvvm", GenericToNVVMPass())
+MODULE_PASS("expand-variadics",
```
[the remark this answers is not in the document]
Comment (jhuber6) [reply llvm/lib/Target/NVPTX/NVPTXPassRegistry.def:20]: Couldn't remember if adding it to `addIRPasses` applied to all uses. I remember something like different pipeline configurations using different things. I'll try to figure it out.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp (hunk 2 above)
Comment (jhuber6) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:964]: So, the standard varargs handling will automatically promote things like shorts to ints and floats to doubles. What the comment means is that `clang` already handled the size / alignment in those cases, so we need to use a minimum alignment of 1 so we respect the alignment for things that clang didn't modify.

**Comment by jhuber6** [comment]
> With the possible exception of some alignment handling this looks about as I'd expect it to. Ideally we'd get some feedback from nvptx-associated people but fixing libc is a good sign

Yep, I believe @Artem-B is on vacation, so hopefully @AlexMaclean can chime in. This should be ABI compatible with NVIDIA as far as I'm aware.

File: clang/lib/Basic/Targets/NVPTX.h [hunk 4]
```diff
@@ -116,8 +116,7 @@ class LLVM_LIBRARY_VISIBILITY NVPTXTargetInfo : public TargetInfo {
   }
 
   BuiltinVaListKind getBuiltinVaListKind() const override {
-    // FIXME: implement
-    return TargetInfo::CharPtrBuiltinVaList;
+    return TargetInfo::VoidPtrBuiltinVaList;
```
Comment (jhuber6) [inline clang/lib/Basic/Targets/NVPTX.h:119]:
```suggestion
    return TargetInfo::CharPtrBuiltinVaList;
```

**Comment by jhuber6** [comment]
Ping, waiting on this so I can land https://github.com/llvm/llvm-project/pull/96369.

**Comment by jlebar** [comment]
@Artem-B

File: clang/lib/CodeGen/Targets/NVPTX.cpp [hunk 5]
```diff
@@ -203,8 +203,12 @@ ABIArgInfo NVPTXABIInfo::classifyArgumentType(QualType Ty) const {
 void NVPTXABIInfo::computeInfo(CGFunctionInfo &FI) const {
   if (!getCXXABI().classifyReturnType(FI))
     FI.getReturnInfo() = classifyReturnType(FI.getReturnType());
+
+  unsigned ArgumentsCount = 0;
   for (auto &I : FI.arguments())
-    I.info = classifyArgumentType(I.type);
+    I.info = ArgumentsCount++ < FI.getNumRequiredArgs()
```
Comment (Artem-B) [inline clang/lib/CodeGen/Targets/NVPTX.cpp:209]: This looks like a good use case for `for (auto &&[index, arg] : llvm::enumerate(FI.arguments)`

**Review by Artem-B — APPROVED** [review]
LGTM with a minor nit.

File: clang/lib/CodeGen/Targets/NVPTX.cpp (hunk 5 above)
Comment (jhuber6) [reply clang/lib/CodeGen/Targets/NVPTX.cpp]: Good point, done.

**Comment by tahonermann** [comment]
The change to `NVPTXTargetInfo::getBuiltinVaListKind()` in `clang/lib/Basic/Targets/NVPTX.h` caused a regression in Intel's downstream Clang-based compiler when compiling SYCL code with device compilation targeting NVPTX. CUDA might be similarly impacted, but I haven't verified. There is no need to revert the change; we've backed out the relevant part of the change in our downstream fork for now. But we need to figure out a proper solution for the problem as described below.

SYCL compilation uses the host standard library headers for both host and device compilation. Microsoft's standard library headers define `va_list` as `char*` as shown at line 72 below:
```
...\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include\vadefs.h:
 67 #ifndef _VA_LIST_DEFINED
 68    #define _VA_LIST_DEFINED
 69    #ifdef _M_CEE_PURE
 70        typedef System::ArgIterator va_list;
 71    #else
 72        typedef char* va_list;
 73    #endif
 74 #endif
 ..
 97     void  __cdecl __va_start(va_list*, ...);
 98     void* __cdecl __va_arg(va_list*, ...);
 99     void  __cdecl __va_end(va_list*);
100 
101     #define __crt_va_start_a(ap, v) ((void)(__va_start(&ap, _ADDRESSOF(v), _SLOTSIZEOF(v), __alignof(v), _ADDRESSOF(v))))
102     #define __crt_va_arg(ap, t)     (*(t *)__va_arg(&ap, _SLOTSIZEOF(t), _APALIGN(t,ap), (t*)0))
103     #define __crt_va_end(ap)        ((void)(__va_end(&ap)))
```

The Clang driver interposes on Microsoft's `vadefs.h` header file by placing its own `vadefs.h` header file earlier in the header search path. This header overrides some of the macros defined by Microsoft's `vadefs.h` header file with ones that are intended to work with Clang's builtin variadic function support.
```
<llvm-project>/clang/lib/Headers/vadefs.h:
 18 #include_next <vadefs.h>
 19
 20 /* Override macros from vadefs.h with definitions that work with Clang. */
 ..
 34 /* VS 2015 switched to double underscore names, which is an improvement, but now
 35  * we have to intercept those names too.
 36  */
 37 #ifdef __crt_va_start
 38 #undef __crt_va_start
 39 #define __crt_va_start(ap, param) __builtin_va_start(ap, param)
 40 #endif
 41 #ifdef __crt_va_end
 42 #undef __crt_va_end
 43 #define __crt_va_end(ap)          __builtin_va_end(ap)
 44 #endif
 45 #ifdef __crt_va_arg
 46 #undef __crt_va_arg
 47 #define __crt_va_arg(ap, type)    __builtin_va_arg(ap, type)
 48 #endif
```

The result is that invocations of the `__crt_va_start`, `__crt_va_end`, and `__crt_va_arg` macros in Microsoft standard library headers end up passing objects of the Microsoft defined `va_list` type (aka, `char*`) to builtin functions like `__builtin_va_start()` that expect objects of type `__builtin_va_list` (aka, `void*`) thus leading to compilation failures when compiling in C++ modes.

There are at least a few options available to try to address this.

1. Revert the change to `NVPTXTargetInfo::getBuiltinVaListKind()` so that `TargetInfo::CharPtrBuiltinVaList` is returned. I don't know what motivated that particular change, so I don't know how feasible this option is.
2. Modify `NVPTXTargetInfo::getBuiltinVaListKind()` to conditionally return a value based on which standard library is being used. I'm not sure how feasible this is either. There are currently two targets that conditionally return different values from their `getBuiltinVaListKind()` implementations; Hexagon (see [here](https://github.com/llvm/llvm-project/blob/8fd9624cf729dd722a170a9dfd8f725966515231/clang/lib/Basic/Targets/Hexagon.h#L106-L110)] and ARM (see [here](https://github.com/llvm/llvm-project/blob/8fd9624cf729dd722a170a9dfd8f725966515231/clang/lib/Basic/Targets/ARM.cpp#L1096-L1101)).
3. Modify Clang's `vadefs.h` header file to also interpose on Microsoft's definition of `va_list` with a change like the following. This could still lead to problems if `_VA_LIST_DEFINED` is already defined, if `va_list` is already declared (as `char*`), or in code that assumes that `va_list` is defined as `char*` in Microsoft environments.
```
 +  #ifndef _VA_LIST_DEFINED
 +      #define _VA_LIST_DEFINED
 +      typedef __builtin_va_list va_list;
 +  #endif
 18 #include_next <vadefs.h>
```

Any thoughts on the above are much appreciated!

**Comment by jhuber6** [comment]
> The change to `NVPTXTargetInfo::getBuiltinVaListKind()` in `clang/lib/Basic/Targets/NVPTX.h` caused a regression in Intel's downstream Clang-based compiler when compiling SYCL code with device compilation targeting NVPTX. CUDA might be similarly impacted, but I haven't verified. There is no need to revert the change; we've backed out the relevant part of the change in our downstream fork for now. But we need to figure out a proper solution for the problem as described below.
> 
> SYCL compilation uses the host standard library headers for both host and device compilation. Microsoft's standard library headers define `va_list` as `char*` as shown at line 72 below:
> 
> ```
> ...\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include\vadefs.h:
>  67 #ifndef _VA_LIST_DEFINED
>  68    #define _VA_LIST_DEFINED
>  69    #ifdef _M_CEE_PURE
>  70        typedef System::ArgIterator va_list;
>  71    #else
>  72        typedef char* va_list;
>  73    #endif
>  74 #endif
>  ..
>  97     void  __cdecl __va_start(va_list*, ...);
>  98     void* __cdecl __va_arg(va_list*, ...);
>  99     void  __cdecl __va_end(va_list*);
> 100 
> 101     #define __crt_va_start_a(ap, v) ((void)(__va_start(&ap, _ADDRESSOF(v), _SLOTSIZEOF(v), __alignof(v), _ADDRESSOF(v))))
> 102     #define __crt_va_arg(ap, t)     (*(t *)__va_arg(&ap, _SLOTSIZEOF(t), _APALIGN(t,ap), (t*)0))
> 103     #define __crt_va_end(ap)        ((void)(__va_end(&ap)))
> ```
> 
> The Clang driver interposes on Microsoft's `vadefs.h` header file by placing its own `vadefs.h` header file earlier in the header search path. This header overrides some of the macros defined by Microsoft's `vadefs.h` header file with ones that are intended to work with Clang's builtin variadic function support.
> 
> ```
> <llvm-project>/clang/lib/Headers/vadefs.h:
>  18 #include_next <vadefs.h>
>  19
>  20 /* Override macros from vadefs.h with definitions that work with Clang. */
>  ..
>  34 /* VS 2015 switched to double underscore names, which is an improvement, but now
>  35  * we have to intercept those names too.
>  36  */
>  37 #ifdef __crt_va_start
>  38 #undef __crt_va_start
>  39 #define __crt_va_start(ap, param) __builtin_va_start(ap, param)
>  40 #endif
>  41 #ifdef __crt_va_end
>  42 #undef __crt_va_end
>  43 #define __crt_va_end(ap)          __builtin_va_end(ap)
>  44 #endif
>  45 #ifdef __crt_va_arg
>  46 #undef __crt_va_arg
>  47 #define __crt_va_arg(ap, type)    __builtin_va_arg(ap, type)
>  48 #endif
> ```
> 
> The result is that invocations of the `__crt_va_start`, `__crt_va_end`, and `__crt_va_arg` macros in Microsoft standard library headers end up passing objects of the Microsoft defined `va_list` type (aka, `char*`) to builtin functions like `__builtin_va_start()` that expect objects of type `__builtin_va_list` (aka, `void*`) thus leading to compilation failures when compiling in C++ modes.
> 
> There are at least a few options available to try to address this.
> 
>     1. Revert the change to `NVPTXTargetInfo::getBuiltinVaListKind()` so that `TargetInfo::CharPtrBuiltinVaList` is returned. I don't know what motivated that particular change, so I don't know how feasible this option is.
> 
>     2. Modify `NVPTXTargetInfo::getBuiltinVaListKind()` to conditionally return a value based on which standard library is being used. I'm not sure how feasible this is either. There are currently two targets that conditionally return different values from their `getBuiltinVaListKind()` implementations; Hexagon (see [here](https://github.com/llvm/llvm-project/blob/8fd9624cf729dd722a170a9dfd8f725966515231/clang/lib/Basic/Targets/Hexagon.h#L106-L110)] and ARM (see [here](https://github.com/llvm/llvm-project/blob/8fd9624cf729dd722a170a9dfd8f725966515231/clang/lib/Basic/Targets/ARM.cpp#L1096-L1101)).
> 
>     3. Modify Clang's `vadefs.h` header file to also interpose on Microsoft's definition of `va_list` with a change like the following. This could still lead to problems if `_VA_LIST_DEFINED` is already defined, if `va_list` is already declared (as `char*`), or in code that assumes that `va_list` is defined as `char*` in Microsoft environments.
> 
> 
> ```
>  +  #ifndef _VA_LIST_DEFINED
>  +      #define _VA_LIST_DEFINED
>  +      typedef __builtin_va_list va_list;
>  +  #endif
>  18 #include_next <vadefs.h>
> ```
> 
> Any thoughts on the above are much appreciated!

Is it just the `char *` vs `void *`? We can probably change that back, in fact I thought I did that in the patch but I might've forgotten to push the update.

**Comment by tahonermann** [comment]
> Is it just the `char *` vs `void *`? We can probably change that back, in fact I thought I did that in the patch but I might've forgotten to push the update.

Yes. The local workaround we're currently using just unconditionally returns `CharPtrBuiltinVaList`.

**Comment by tahonermann** [comment]
For historical reference, @jhuber6 backed out the change to `NVPTXTargetInfo::getBuiltinVaListKind()` in https://github.com/llvm/llvm-project/pull/100438 (commit https://github.com/llvm/llvm-project/commit/fb1e077982e90e715058188fa667ea5dd8027a98)

**Pull request merged**